### DISCUSSION TITLE
Nedelec element formulation

### DISCUSSION THREAD

**Discussion by karthichockalingam** [idaholab/moose#25324] [Q&A General] [answered]
Hello,

@nmnobre and I trying to understand the Nedelec implementation.  In particular [feCurlPhi](https://github.com/idaholab/moose/blob/5473cf3dc207ee0c6d80ed5bf0cc4cda295a5eb8/framework/src/base/Assembly.C#L4643) has the return type ```VariablePhiCurl``` whereas ```_vector_fe_shape_data[type]->_curl_phi ``` is of type ```VectorVariablePhiCurl``` as defined [here](https://github.com/idaholab/moose/blob/5473cf3dc207ee0c6d80ed5bf0cc4cda295a5eb8/framework/include/base/Assembly.h#L2624C31-L2624C31). So shouldn't the return function type be ```VectorVariablePhiCurl``` instead? 

But having the return type ```VectorVariablePhiCurl``` would lead to compilation error as it is a specialisation of the [template](https://github.com/idaholab/moose/blob/5473cf3dc207ee0c6d80ed5bf0cc4cda295a5eb8/framework/include/base/Assembly.h#L1641)

I was under the impression Nedelec implementation is for vector-valued problems. Why is the return type of ```feCurlPhi``` not ```VectorVariablePhiCurl```?

Many thanks for the clarification!

Karthi

**Comment by GiudGiud** [comment]
@cticenhour

**Comment by lindsayad** [comment, marked as answer]
`feCurlPhi` is a template
```c++
  template <typename OutputType>
  const typename OutputTools<OutputType>::VariablePhiCurl & feCurlPhi(FEType type) const
  {
    _need_curl[type] = true;
    buildFE(type);
    return _fe_shape_data[type]->_curl_phi;
  }
```
You can see the vector specialization:
```c++
template <>
const typename OutputTools<VectorValue<Real>>::VariablePhiCurl &
Assembly::feCurlPhi<VectorValue<Real>>(FEType type) const
{
  _need_curl[type] = true;
  buildVectorFE(type);
  return _vector_fe_shape_data[type]->_curl_phi;
}
```
curl is not *a priori* tied to Nedelece elements. You can implement a curl for `LAGRANGE_VEC` as well for example

**Reply by lindsayad** [reply]
Having curl available for scalar finite element types allows us to reduce the number of routine implementations, such as for `MooseVariableData::computeValues`

**Reply by karthichockalingam** [reply]
I can see, you want to avoid code duplication when dealing with scalar-valued problems.
However, I don't see how it works out for vector-valued problems. 

Why is the return type ```const typename OutputTools<VectorValue<Real>>::VariablePhiCurl``` because ```Phi``` is clearly a vector. Shouldn't the return type be ```VectorVariablePhiCurl``` instead?

Yes, thanks for pointing out to me that curl can be computed in other spaces too.

**Reply by lindsayad** [reply]
> However, I don't see how it works out for vector-valued problems.

Why not? We get to use the exact same `computeValues` routine for all finite element types. Code duplication was the wrong phrase to use perhaps, which is why I edited the comment to say "number of routine implementations". The latter is important to stop cases where a developer implements a new feature or fixes a bug in the general template but forgets to add it in a specialization (which would then be an example of code duplication when they then go and add the feature for the specialization)

> Why is the return type const typename OutputTools<VectorValue<Real>>::VariablePhiCurl because Phi is clearly a vector. Shouldn't the return type be VectorVariablePhiCurl instead?

This is from `MooseTypes.h`
```c++
template <typename OutputType>
struct OutputTools
{
  typedef typename TensorTools::IncrementRank<OutputType>::type OutputGradient;
  typedef typename TensorTools::IncrementRank<OutputGradient>::type OutputSecond;
  typedef typename TensorTools::DecrementRank<OutputType>::type OutputDivergence;

  typedef MooseArray<OutputType> VariableValue;
  typedef MooseArray<OutputGradient> VariableGradient;
  typedef MooseArray<OutputSecond> VariableSecond;
  typedef MooseArray<OutputType> VariableCurl;
  typedef MooseArray<OutputDivergence> VariableDivergence;

  typedef typename Moose::ShapeType<OutputType>::type OutputShape;
  typedef typename TensorTools::IncrementRank<OutputShape>::type OutputShapeGradient;
  typedef typename TensorTools::IncrementRank<OutputShapeGradient>::type OutputShapeSecond;
  typedef typename TensorTools::DecrementRank<OutputShape>::type OutputShapeDivergence;

  typedef MooseArray<std::vector<OutputShape>> VariablePhiValue;
  typedef MooseArray<std::vector<OutputShapeGradient>> VariablePhiGradient;
  typedef MooseArray<std::vector<OutputShapeSecond>> VariablePhiSecond;
  typedef MooseArray<std::vector<OutputShape>> VariablePhiCurl;
  typedef MooseArray<std::vector<OutputShapeDivergence>> VariablePhiDivergence;

  typedef MooseArray<std::vector<OutputShape>> VariableTestValue;
  typedef MooseArray<std::vector<OutputShapeGradient>> VariableTestGradient;
  typedef MooseArray<std::vector<OutputShapeSecond>> VariableTestSecond;
  typedef MooseArray<std::vector<OutputShape>> VariableTestCurl;
  typedef MooseArray<std::vector<OutputShapeDivergence>> VariableTestDivergence;

  // DoF value type for the template class OutputType                                                                 
  typedef typename Moose::DOFType<OutputType>::type OutputData;
  typedef MooseArray<OutputData> DoFValue;
  typedef OutputType OutputValue;
};
```
Instantiating `OutputTools<VectorValue<Real>>` results in `typename OutputTools<VectorValue<Real>>::VariablePhiCurl` corresponding to a `MooseArray<std::vector<VectorValue<Real>>>` which is the equivalent of the `VectorVariablePhiCurl`. This is the `typedef` in `MooseTypes.h`:
```c++
typedef typename OutputTools<RealVectorValue>::VariablePhiCurl VectorVariablePhiCurl;
```

**Reply by karthichockalingam** [reply]
Thank you very much for your response. This clarifies my question.
I got confused because I thought the following type would be instantiated instead:

``` typedef typename OutputTools<RealVectorValue>::VariablePhiCurl VectorVariablePhiCurl;``` 

You are in way a telling me that types ```<RealVectorValue>``` and ```<VectorValue<Real>>``` are not the same. I guess that was the whole purpose of the developer all along prototyping a general template. 

We are looking into implementing RT elements for vector-valued problems. So 

```OutputTools<OutputType>::VariablePhiDivergence ``` 

wouldn't through up an error for ```Real``` types (or is there divergence for scalar field)? Do templates in general get initiated for all types or only the specific case it is invoked for (i.e. VectorValue<Real>)?

**Reply by lindsayad** [reply]
`VectorValue<Real>` and `RealVectorValue` are actually the same. I'm sure that causes some confusion and I'm sorry about that!

> We are looking into implementing RT elements for vector-valued problems

That would be awesome. I would love to see that. Have you looked at libMesh which is our underlying finite element library? That's actually the place where a new basis implementation would go and then it would propagate down into MOOSE

> wouldn't through up an error for Real types (or is there divergence for scalar field)?

No it would not. There is not really a divergence for a scalar field. We could consider doing some refactoring in our variable classes that remove curl and divergence code for all but the vector template instances. However, we just haven't had a need motivating that change.

> Do templates in general get initiated for all types or only the specific case it is invoked for (i.e. VectorValue)?

They are only instantiated when used (implicitly) or when explicitly instantiated. For instance at the bottom of `MooseVariableFE.C` we perform some explicit instantiations:
```c++
template class MooseVariableFE<Real>;
template class MooseVariableFE<RealVectorValue>;
template class MooseVariableFE<RealEigenVector>;
```

**Reply by karthichockalingam** [reply]
Yes, we have completed the RT implementation in libmesh, testing is underway to see if we get the right convergence. We just started looking to port it into MOOSE. I am sure, we will have questions along the way.

**Reply by lindsayad** [reply]
Very exciting stuff. Please continue asking questions as needed!

**Reply by karthichockalingam** [reply]
Is there a way to print the system matrix before it is sent to the solver?
If so, where can I add that line in the framework?

**Reply by GiudGiud** [reply]
Are you using PJNK or Newton?
-ksp_view_pmat on the command line will show the preconditioning matrix. In general the petsc command line options can be passed on the command line in moose (or in the executioner block) to get the petsc output

**Reply by lindsayad** [reply]
Yes @GiudGiud’s command will print the system matrix. For PJFNK we use the system matrix for P (matrix for creating a preconditioner for the linearized Jacobian A) and for NEWTON we use it for both A and P
